File: scripts/maintenance/localize_scripts.py
import os
import re
import hashlib
import subprocess
import random
import time
from pathlib import Path
from urllib.parse import urlparse

import sys

SCRIPTS_DIR = Path(__file__).resolve().parents[1]
sys.path.insert(0, str(SCRIPTS_DIR))
from lib.common import _BROWSER_UA, safe_download_binary, Logger

ROOT = Path(__file__).parent.parent.parent
MODULE_DIRS = [ROOT / "module/surge(main)", ROOT / "module/shadowrocket"]
SCRIPTS_DIR = ROOT / "module/scripts"
LOCAL_URL_PREFIX = "https://raw.githubusercontent.com/nowaytouse/script_hub/master/module/scripts/"

# User-Agent Pool
USER_AGENTS = [_BROWSER_UA]

def is_valid_url(url):
    """Filter out regex patterns or invalid mock URLs."""
    if ".*" in url or "{{" in url or url.endswith(".js*"):
        return False
    parsed = urlparse(url)
    if not parsed.netloc or "." not in parsed.netloc:
        return False
    return True

def download_script(url):
    """Download script with centralized download utility."""
    if not is_valid_url(url):
        print(f"  ⏭️  Skipping invalid/regex URL: {url}")
        return None

    parsed = urlparse(url)
    domain = parsed.netloc.replace(".", "_")
    path_parts = parsed.path.strip("/").split("/")
    filename = path_parts[-1]
    if not filename.endswith(".js"):
        filename += ".js"
    
    path_hash = hashlib.md5(parsed.path.encode()).hexdigest()[:6]
    local_filename = f"{domain}_{path_hash}_{filename}"
    local_path = SCRIPTS_DIR / local_filename
    
    # If already downloaded, no need to re-download unless it's a force sync
    if local_path.exists():
        return local_filename

    print(f"  📥 Downloading script: {url}")
    content = safe_download_binary(url, retries=2, timeout=15)
    if content:
        with open(local_path, "wb") as f:
            f.write(content)
        return local_filename
    
    print(f"  ⚠️  Keep remote (download failed): {url}")
    return None

def localize_module(file_path):
    with open(file_path, "r", encoding="utf-8") as f:
        content = f.read()
    
    urls = re.findall(r'https?://[^\s,]+?\.js', content)
    if not urls:
        return

    new_content = content
    changed = False
    
    for url in set(urls):
        if LOCAL_URL_PREFIX in url:
            continue
            
        local_filename = download_script(url)
        if local_filename:
            local_url = LOCAL_URL_PREFIX + local_filename
            new_content = new_content.replace(url, local_url)
            changed = True
    
    if changed:
        with open(file_path, "w", encoding="utf-8") as f:
            f.write(new_content)
        print(f"🔍 Updated: {file_path.relative_to(ROOT)}")

def main():
    if not SCRIPTS_DIR.exists():
        os.makedirs(SCRIPTS_DIR)
        
    for root_dir in MODULE_DIRS:
        if not root_dir.exists():
            continue
        for ext in ["*.sgmodule", "*.module"]:
            for module_file in root_dir.rglob(ext):
                localize_module(module_file)

if __name__ == "__main__":
    main()

File: create/tools/localize_scripts.go
package tools

import (
	"crypto/md5"
	"encoding/hex"
	"fmt"
	"net/url"
	"os"
	"path/filepath"
	"regexp"
	"strings"

	"github.com/nyamiiko/script_hub/create/hub"
)

var (
	skipDomains = map[string]bool{
		"cdn.jsdelivr.net": true,
	}
)

func isValidUrl(u string) bool {
	if strings.Contains(u, ".*") || strings.Contains(u, "{{") || strings.HasSuffix(u, ".js*") {
		return false
	}
	parsed, err := url.Parse(u)
	if err != nil || parsed.Host == "" || !strings.Contains(parsed.Host, ".") {
		return false
	}
	if skipDomains[parsed.Host] {
		return false
	}
	return true
}

func downloadScript(u string) string {
	if !isValidUrl(u) {
		fmt.Printf("  ⏭️  Skipping invalid/regex URL: %s\n", u)
		return ""
	}

	parsed, _ := url.Parse(u)
	domain := strings.ReplaceAll(parsed.Host, ".", "_")
	pathParts := strings.Split(strings.Trim(parsed.Path, "/"), "/")
	filename := pathParts[len(pathParts)-1]
	if !strings.HasSuffix(filename, ".js") {
		filename += ".js"
	}

	hash := md5.Sum([]byte(parsed.Path))
	pathHash := hex.EncodeToString(hash[:])[:6]
	localFilename := fmt.Sprintf("%s_%s_%s", domain, pathHash, filename)
	localPath := filepath.Join(hub.MODULE_SOURCE_SCRIPTS_DIR, localFilename)

	if hub.ValidateFileExists(localPath, "") {
		return localFilename
	}

	fmt.Printf("  📥 Downloading script: %s\n", u)
	content := hub.SafeDownload(u, 2, 15)
	if content != "" {
		hub.SafeWriteFile(localPath, content, true)
		return localFilename
	}

	fmt.Printf("  ⚠️  Keep remote (download failed): %s\n", u)
	return ""
}

func localizeModule(filePath string) {
	content := hub.ReadFileString(filePath)
	if content == "" {
		return
	}

	urls := regexp.MustCompile(`script-path=([^,\s]+)`).FindAllStringSubmatch(content, -1)
	if len(urls) == 0 {
		return
	}

	newContent := content
	changed := false

	uniqueUrls := make(map[string]bool)
	for _, m := range urls {
		uniqueUrls[m[1]] = true
	}

	for u := range uniqueUrls {
		if strings.Contains(u, hub.SCRIPT_RAW_PREFIX) {
			continue
		}

		localFilename := downloadScript(u)
		if localFilename != "" {
			localUrl := hub.SCRIPT_RAW_PREFIX + localFilename
			newContent = strings.ReplaceAll(newContent, u, localUrl)
			changed = true
		}
	}

	if changed {
		hub.SafeWriteFile(filePath, newContent, true)
		rel, _ := filepath.Rel(hub.ROOT, filePath)
		fmt.Printf("🔍 Updated: %s\n", rel)
	}
}

func RunLocalizeScripts() int {
	hub.EnsureDir(hub.MODULE_SOURCE_SCRIPTS_DIR)

	moduleDirs := []string{
		filepath.Join(hub.ROOT, "modules/surge"),
		filepath.Join(hub.ROOT, "modules/shadowrocket"),
		filepath.Join(hub.ROOT, "modules/source/local"),
		filepath.Join(hub.ROOT, "rulesets/Sources/LocalModules"),
	}

	for _, rootDir := range moduleDirs {
		if !hub.ValidateDirExists(rootDir, "") {
			continue
		}

		filepath.Walk(rootDir, func(path string, info os.FileInfo, err error) error {
			if err != nil {
				return nil
			}
			if !info.IsDir() && (strings.HasSuffix(info.Name(), ".sgmodule") || strings.HasSuffix(info.Name(), ".module")) {
				localizeModule(path)
			}
			return nil
		})
	}
	return 0
}
